### DISCUSSION TITLE
Escaping callback closures execution order

### DISCUSSION THREAD

**Discussion by kublaios** [Tyler-Keith-Thompson/CucumberSwift#51] [Q&A] [answered]
In my use case, I want to access a variable that's assigned during `Given`/`When` steps in `Then` step. Checking the callback closure in `Given`/`When` initializers, I see that they are `escaping`. I want to understand the risks when I assume that the closures will run serially like my example below.

```
struct MyFeature {
    static func setupSteps(app: XCUIApplication) {
        Given("^The user is on home screen$") { _, _ in
            XCTAssertTrue(app.staticTexts["Welcome!"].exists)
        }
        var mapScreen: MapScreen!
        When("^Map button is tapped$") { _, _ in
            mapScreen = HomeScreen(app: app).tapMapButton()
        }
        Then("^Map is visible$") { _, _ in
            XCTAssertTrue(mapScreen.verifyMapViewVisible())
        }
    }
}
```

Note that this code works as expected, meaning I have `mapScreen` assigned by the time the execution reaches the line I use it.

**Comment by Tyler-Keith-Thompson** [comment, marked as answer]
The guarantee the library makes is that these will execute in the order defined in your `.feature` file. That means I could crash your tests by writing:

```gherkin
Given The user is on home screen
Then Map is visible
```

However, unless I maliciously manipulate your Gherkin (or you write such generic steps that accidental re-use causes issues) your code should be safe.

One way you could choose to rewrite this so that it fails, but doesn't crash is:

```swift
Given("^The user is on home screen$") { _, _ in
    XCTAssertTrue(app.staticTexts["Welcome!"].exists)
}
var mapScreen: MapScreen?
When("^Map button is tapped$") { _, _ in
    mapScreen = HomeScreen(app: app).tapMapButton()
}
Then("^Map is visible$") { _, _ in
    let screen = try XCTUnwrap(mapScreen) // any errors thrown will get caught by XCTest
    XCTAssertTrue(screen.verifyMapViewVisible())
}
```

**Reply by Tyler-Keith-Thompson** [reply]
One interesting note: the library doesn't guarantee thread safety. I'll be adding async closures to steps at some point. Really all that means is you're responsible for your fate in that regard. So you could also crash this simply by wrapping your assignment to `mapScreen` in a background thread. That'd cause a race condition.

**Reply by kublaios** [reply]
Thanks, that's all I needed to know.
